chore: add migration doc for nodejs (#164)
elastic/opentelemetry: docs/_edot-sdks/nodejs/migration.md

For example: `OTEL_RESOURCE_ATTRIBUTES=service.version=1.2.3`.

### `environment`

The Elastic [`environment`](https://www.elastic.co/guide/en/apm/agent/nodejs/current/configuration.html#environment) option corresponds to setting the `deployment.environment` key in [OTEL_RESOURCE_ATTRIBUTES](https://opentelemetry.io/docs/concepts/sdk-configuration/general-sdk-configuration/#otel_resource_attributes).

For example: `OTEL_RESOURCE_ATTRIBUTES=deployment.environment=testing`.

### `globalLabels`

The Elastic [`globalLabels`](https://www.elastic.co/guide/en/apm/agent/nodejs/current/configuration.html#global-labels) option corresponds to adding `key=value` comma separated pairs in [OTEL_RESOURCE_ATTRIBUTES](https://opentelemetry.io/docs/concepts/sdk-configuration/general-sdk-configuration/#otel_resource_attributes).

For example: `OTEL_RESOURCE_ATTRIBUTES=alice=first,bob=second`. Such labels will result in labels.key=value attributes on the server, e.g. labels.alice=first

### `transactionSampleRate`

> The Elastic [`transactionSampleRate`](https://www.elastic.co/guide/en/apm/agent/nodejs/current/configuration.html#transaction-sample-rate) option does not directly correspond to a OpenTelemetry option but the same behaviour can be achieved using `OTEL_TRACES_SAMPLER` and `OTEL_TRACES_SAMPLER_ARG` options. OpenTelemetry Node.js SDK comes with the built-in `TraceIdRatioBased` sampler which accepts an argument for the sample rate.

Reviewer: nit: 
```suggestion
The Elastic [`transactionSampleRate`](https://www.elastic.co/guide/en/apm/agent/nodejs/current/configuration.html#transaction-sample-rate) option does not directly correspond to an OpenTelemetry option but the same behavior can be achieved using `OTEL_TRACES_SAMPLER` and `OTEL_TRACES_SAMPLER_ARG` options. OpenTelemetry Node.js SDK comes with the built-in `TraceIdRatioBased` sampler which accepts an argument for the sample rate.
```

# Migrating to EDOT Node.js from the Elastic Node.js Agent

This documentation describes how to update applications that are currently using the [Elastic APM Node.js agent](https://www.elastic.co/guide/en/apm/agent/nodejs/current/index.html) to use the Elastic Distribution of OpenTelemetry for Node.js (EDOT Node.js).

## Advantages of using EDOT Ndode.js agent

### OpenTelemetry-native Data

Allows to capture, send, transform and store data in an OpenTelemetry native way. This includes for example the ability to use all features of the OpenTelemetry SDK for manual tracing, data following semantic conventions or ability to use intermediate collectors and processors.

> 

Author: I'll refer to `@opentelemetry/auto-instrumentations-node` instead of "upstream agent"

## Limitations

### Missing Instrumentations

> EDOT Node.js only supports a technologies listed in [Supported Technologies](./supported-technologies) page. Although they do not appear in the list `AWS Lambda` and `Azure Functions` can be instrumented other means

Author: thanks for the rewording :)

For example: `OTEL_RESOURCE_ATTRIBUTES=host.name=my-host`.

### `logLevel`

The Elastic [`logLevel`](https://www.elastic.co/guide/en/apm/agent/nodejs/current/configuration.html#log-level) option corresponds to the OpenTelemetry [`OTEL_LOG_LEVEL`](https://opentelemetry.io/docs/specs/otel/configuration/sdk-environment-variables/#general-sdk-configuration) option. The possible values change a bit but they represent similar levels. The following
table shows the equivalent values of log levels between `elastic-apm-node` and EDOT Node.js

| ELASTIC_APM_LOG_LEVEL | OTEL_LOG_LEVEL |
| --------------------- | -------------- |
| `off`                 | `none`         |
| `error`               | `error`        |
| `warn`                | `warn`         |
| `info`                | `info`         |
| `debug`               | `debug`        |
| `trace`               | `verbose`      |
| `trace`               | `all`          |

### `maxQueueSize`

> The Elastic [`maxQueueSize`](https://www.elastic.co/guide/en/apm/agent/nodejs/current/configuration.html#max-queue-size) option corresponds to the OpenTelemetry [`OTEL_BSP_MAX_QUEUE_SIZE`](https://opentelemetry.io/docs/specs/otel/configuration/sdk-environment-variables/#batch-span-processor) option.

Author: Haven't found the equiv for metrics in the docs and neither in the code

For example: `OTEL_BSP_MAX_QUEUE_SIZE=4096`.

### `serverTimeout`

The Elastic [`serverTimeout`](https://www.elastic.co/guide/en/apm/agent/nodejs/current/configuration.html#server-timeout) option corresponds to the OpenTelemetry [`OTEL_BLRP_EXPORT_TIMEOUT`](https://opentelemetry.io/docs/specs/otel/configuration/sdk-environment-variables/#batch-span-processor) option.

> For example: `OTEL_BLRP_EXPORT_TIMEOUT=50000`.

Reviewer: Similarly, this section should doc both this one and the `OTEL_BSP_EXPORT_TIMEOUT` for exporting spans.

Author: and for metrics too :)

### `apmClientHeaders`

The Elastic [`apmClientHeaders`](https://www.elastic.co/guide/en/apm/agent/nodejs/current/configuration.html#apm-client-headers) option corresponds to the OpenTelemetry [`OTEL_EXPORTER_OTLP_HEADERS`](https://opentelemetry.io/docs/specs/otel/protocol/exporter/#specifying-headers-via-environment-variables) option.

For example: `OTEL_EXPORTER_OTLP_HEADERS=foo=bar,baz=quux`.

### `disableInstrumentations`

> The Elastic [`disableInstrumentations`](https://www.elastic.co/guide/en/apm/agent/nodejs/current/configuration.html#apm-client-headers) option corresponds to the OpenTelemetry [`OTEL_NODE_DISABLED_INSTRUMENTATIONS`](https://opentelemetry.io/docs/zero-code/js/configuration/#disable-specific-instrumentations) option. See [Configuration](./configuration) if you want to learn more advanced usage.

Reviewer: ```suggestion
The Elastic [`disableInstrumentations`](https://www.elastic.co/guide/en/apm/agent/nodejs/current/configuration.html#apm-client-headers) option corresponds to the EDOT Node.js [`OTEL_NODE_DISABLED_INSTRUMENTATIONS`](https://elastic.github.io/opentelemetry/edot-sdks/nodejs/configuration.html#otel_node_disabledenabled_instrumentations-details) option.
```

For example: `OTEL_RESOURCE_ATTRIBUTES=container.id=my-id`.

### `metricsInterval`

The Elastic [`metricsInterval`](https://www.elastic.co/guide/en/apm/agent/nodejs/current/configuration.html#metrics-interval) option corresponds to the OpenTelemetry [`OTEL_METRIC_EXPORT_INTERVAL`](https://opentelemetry.io/docs/specs/otel/configuration/sdk-environment-variables/#periodic-exporting-metricreader) option.

For example: `OTEL_METRIC_EXPORT_INTERVAL=30000`.

### `cloudProvider`

The Elastic [`cloudProvider`](https://www.elastic.co/guide/en/apm/agent/nodejs/current/configuration.html#cloud-provider) option does not corresponds directly to an OpenTelemetry option but you can get similar behaviour by properly setting [`OTEL_NODE_RESOURCE_DETECTORS`](https://opentelemetry.io/docs/zero-code/js/configuration/#sdk-resource-detector-configuration) option. You have to make sure the
cloud provider is on the list of detectors. Not setting this option is the equivalent of `auto`.

> For example: `OTEL_NODE_RESOURCE_DETECTORS=os,env,host,process,gcp`. Will make the agent query for GCP metadata only.

Reviewer: This is missing some non-cloud detectors, I think. `serviceinstance` and `container`. Could refer to https://elastic.github.io/opentelemetry/edot-sdks/nodejs/configuration.html#otel_node_resource_detectors-details instead of the opentelemetry.io link.

TODO:
https://www.elastic.co/guide/en/apm/agent/nodejs/current/configuration.html#long-field-max-length

is it maybe attr limits on each signal?
https://opentelemetry.io/docs/specs/otel/configuration/sdk-environment-variables/#attribute-limits
https://opentelemetry.io/docs/specs/otel/configuration/sdk-environment-variables/#span-limits

> https://opentelemetry.io/docs/specs/otel/configuration/sdk-environment-variables/#logrecord-limits

Reviewer: Yes, just the `attribute-limits` link will suffice. Users can use the span or logrecord specific ones if they *really* need to.

It *might* be worthwhile to defer handling this config var to a separate issue to finish documenting all the config vars. The subtlety with attribute limits is that the downstream Elasticsearch might effectively truncate (or at least limit indexing to the first 1000 chars) no matter how the SDK is configured.
